### PR TITLE
Feature/gbr 35

### PR DESCRIPTION
**Pull request by mikeborozdin**
Changed

### CODE REVIEW

**Comment by mikeborozdin** [comment]
It's for https://github.com/caitp/grunt-bower-release/issues/35

File: tasks/bower-release.js [hunk 1]
```diff
@@ -41,7 +41,7 @@ module.exports = function(grunt) {
    * falsy ones) are ignored.
    */
   var endpoints = {
-    git: require('./endpoints/git')(grunt),
+    git: require('./endpoints/git')(grunt, require('async')),
```
Comment (caitp) [inline tasks/bower-release.js]: Ordinarily, this is available via `grunt.util.async`, so it's not necessarily needed here.

That said, the version bundled with grunt is deprecated and may be removed in the future, so it's fine to provide a separate copy.

However I'd prefer to avoid including it this way, better to just require it from within the endpoint itself, or alternatively use a solution like https://github.com/angular/di.js to manage the dependencies (which would be useful anyways because it should help with testing).

File: tasks/bower-release.js [hunk 2]
```diff
@@ -263,17 +263,39 @@ module.exports = function(grunt) {
...
         endpoint.commit(msg, function(err) {
-          /* Tag name must be valid semver -- but I'm not validating this here. */
-          /* TODO: Validate this here! */
-          endpoint.tag(bowerJSON.version, makeTagMsg(options.packageName), tagged)
+          if (options.overwriteVersion === true) {
+            endpoint.getVersionTags(bowerJSON.version, gotVersionTags);
+          } else {
+            /* Tag name must be valid semver -- but I'm not validating this here. */
+            /* TODO: Validate this here! */
+            console.log('hello');
```
Comment (caitp) [inline tasks/bower-release.js]: lets remove this

**Comment by caitp** [comment]
There are a few comments left that should be easy to fix, I'll give it another look when those are cleaned up.

Please don't bump the version number in package.json as this is automated --- and when you hear LGTM from me, I would appreciate that all of these commits are squashed with a meaningful commit message logging all of the changes, since this is a fairly substantial feature/fixup.

File: tasks/bower-release.js (hunk 1 above)
Comment (mikeborozdin) [inline tasks/bower-release.js]: Hi,

I am using `async` exactly because grunt.util.async is deprecated. Sorry, I left a comment about it in the previous pull request, but forgot to put it here.

The reason I am injecting it, albeit manually, is because I have a test for it in https://github.com/NCR-CoDE/grunt-bower-release/blob/feature/gbr-35/test/unit/tasks/endpoints/git-test.js#L3

Thanks for telling me about Angular 2.0 DI. Unfortunately, it is not on NPM yet, hence it will difficult to include it as a dependency. There is though a previous iteration of it (https://www.npmjs.org/package/di), however, as it GitHub (https://github.com/vojtajina/node-di) page says it's been deprecated in favour of the aforementioned Angular 2.0 DI.

The page claims they are similar, however, I can see some differences, as Angular 2.0 DI doesn't seem to use Guice inspired modules, while Node DI is doing that a lot.

Let me know, if you still want to see `node di` in here of if we can file a TODO item when Angular 2.0 DI hits NPM.

Comment (caitp) [inline tasks/bower-release.js]: You can add di.js to package.json as a dependency using its git url, for instance

``` js
"di": "git://github.com/angular/di.js#master"
```

Should work

Comment (mikeborozdin) [inline tasks/bower-release.js]: Thanks! It sounds like a plan!

Are you okay though with binding it to the master branch? Potentially, there can be a breaking change that will stop grunt-bower-release from working. And there is no tag in their Git repo to refer to, so that at least we depend on the same version.

Comment (mikeborozdin) [inline tasks/bower-release.js]: Having said that, we can bind our dependency to the current commit in order to make sure that new changes will not break out code.

Comment (caitp) [inline tasks/bower-release.js]: I'm on the angular core team, so I will hopefully be aware of breaking changes. they've not been happening much recently.

However, Karma is still using "di", and it might not hurt to just stick with that due to stability until there's a formal release

Comment (mikeborozdin) [inline tasks/bower-release.js]: @caitp,

I don't think it will work. Even though we are able to get the node dependency from GitHub, it has to be transpiled from ES6 to ES5.

There is a possible workaround that we provide a `postinstall script` which first installs `gulp` and then uses it to compile. However, I think it's too hacky...

Comment (caitp) [inline tasks/bower-release.js]: di.js doesn't need to be transpiled from es6. There is an ES5 cjs release, and traceur annotations aren't required

Comment (mikeborozdin) [inline tasks/bower-release.js]: @caitp,

Thanks a lot for your help!

But I think I'm missing something - where is that release? If I look at [node/index.js](https://github.com/angular/di.js/blob/master/node/index.js), then it refers to non-existent `.../dist/cjs/index.js` that should be produced by gulp...

Comment (caitp) [inline tasks/bower-release.js]: Yeah, most of our v2 repositories have a dist branch with precompiled code, I'll push one to di.js later on.

HOWEVER

IMO, just go with node-di, Karma uses it and it works fairly nicely.

Alternatively, could just require async directly from the endpoint, and use a different strategy to mock it for tests ("mockery" is one that's good for this)

**Comment by mikeborozdin** [comment]
@caitp,

I have pushed the changed you requested.

I've opted out for using node-di. While `mockery` would be easy to use, it seemed to be too implicit, while dependencies should be more explicit. 

I still have to pass an instance of grunt to an injection module, but it's because we are writing a grunt plug-in and there's no other way to get hold of a grunt instance.

Please, note that in git-test.js I am still directly injecting my mocks, but that is fine for that particular purpose.

File: tasks/bower-release.js [hunk 3]
```diff
@@ -25,6 +25,14 @@
 'use strict';
 
 module.exports = function(grunt) {
+
+  var di = require('di');
+  var injectionModule = require('./injection-module');
```
Comment (caitp) [inline tasks/bower-release.js:30]: Just call this Module, tbh

Comment (mikeborozdin) [inline tasks/bower-release.js:30]: You think so? I find `injection-module` to be more explicit, especially when just looking at the file name without knowing the entire context.

File: tasks/bower-release.js [hunk 4]
```diff
@@ -25,6 +25,14 @@
...
+  var injector = new di.Injector([injectionModule(grunt)]);
```
Comment (caitp) [inline tasks/bower-release.js:31]: the Module constructor doesn't actually take any parameters

Comment (mikeborozdin) [inline tasks/bower-release.js:31]: How would we pass an instance of grunt otherwise? 

We have to pass it to git, so that it can use `grunt.util.spawn` and legacy util.*

Comment (caitp) [inline tasks/bower-release.js:31]: You can instead do this:

``` js
var module = new Module();
module.
  value("grunt", grunt).
  value("async", async);
// ...

var injector = new di.Injector(module);
/// ...
```

Comment (caitp) [inline tasks/bower-release.js:31]: You can also say:

``` js
var module = {
  "grunt": [ "grunt", "value", grunt ],
  "async": [ "async", "value", async ]
};
```

but I like the syntax-sugar'd version better

Comment (mikeborozdin) [inline tasks/bower-release.js:31]: I can. However, how about trying to split up a module (where you can modify dependencies) from a code that is using that module to create an injector? After all, node-di resembles [Guice](https://code.google.com/p/google-guice/wiki/GettingStarted) pretty much where you always keep modules separately.

Comment (caitp) [inline tasks/bower-release.js:31]: I'm all for it.

The task needs to add grunt to the module before creating the inject, if and only if no mock is present. A mock can be detected using module.forEach()

Comment (mikeborozdin) [inline tasks/bower-release.js:31]: Why if and only if? Will it be any different if we are declaring a module inside the task? You still rely on an instance of grunt that is passed inside the task.

Comment (caitp) [inline tasks/bower-release.js:31]: if `grunt` is mocked (which is unlikely, but possible), you don't want to add the proper `grunt` instance to the injector, that's all I'm saying.

Comment (mikeborozdin) [inline tasks/bower-release.js:31]: Ah! I think I see what you mean right now!

Are you saying that the proper instance can be injected in my injectionModule? But how? I am calling it inside the task, while it is the task that receives an instance of `grunt` whether it is a real one or mocked one.

Yes, both of those declarations look like:

```
exports.module = function(grunt) {}
```

However, isn't it only bower-release.json that is instantiated by grunt and where it passes a reference to itself?

Comment (caitp) [inline tasks/bower-release.js:31]: maybe it's a bit much to hack on an IOC container for this in this change, maybe we can keep it like this for now and I'll work on adding an IOC container seperately, unless you're interested in adding that (but maybe it should still be in a separate CL, just to make this one easier to review).

It's true that grunt makes things a little bit awkward, but it's not all bad. You can have a file, like `modules.js` which exports a function that creates the default modules, and define an entrypoint for mocking default modules there. The grunt task itself can create the injector by requiring those modules, and adding the `grunt` provider if it's not determined to be already present in the module, so that a mock is not overwritten.

Does that make sense? It should be a pretty simple architecture, but it's true that Grunt makes it a bit awkward to build that.

Comment (mikeborozdin) [inline tasks/bower-release.js:31]: Well, I think we can think about it in a little bit more details later. I also suppose we will have feature requests for keeping up-to-date with our dependencies, i.e. replacing `grunt.util.async` with `async` and node-di with `di.js` when it is ready.

Just let me know if there is anything else to be done on this pull request and I'll be happy to squash my commits.

Comment (caitp) [inline tasks/bower-release.js:31]: I'm going to be unavailable for most of tomorrow, but I'll try to give this another look on Thursday, SGTY?

A couple of things which do come up: mind the EOL characters in files altered in the patch, they add nonsense to the diff which doesn't need to be there. You can correct these by grabbing https://github.com/caitp/grunt-bower-release/pull/40.patch and removing the lines which don't seem to alter anything except changing the EOL marker, and then applying the patch by checking out the master branch and running `cat <thepatchname>.patch | git am` (probably the easiest way to fix that that I'm aware of). I'll try to leave some other comments here real quick if anything else comes to mind quickly

File: tasks/bower-release.js [hunk 5]
```diff
@@ -25,6 +25,14 @@
...
+
+  var Git = require('./endpoints/git');
+  var Test = require('./endpoints/test-ep');
```
Comment (caitp) [inline tasks/bower-release.js:34]: why are we requiring `test-ep` outside of an actual test?

File: tasks/bower-release.js [hunk 6]
```diff
@@ -41,8 +49,8 @@ module.exports = function(grunt) {
    * falsy ones) are ignored.
    */
   var endpoints = {
-    git: require('./endpoints/git')(grunt),
-    test: require('./endpoints/test-ep')(grunt)
+    git: injector.instantiate(Git),
+    test: injector.instantiate(Test)
   }
```
Comment (caitp) [inline tasks/bower-release.js:54]: If you don't remove the IOC container stuff from this PR, it might be a good idea to add this as a value in the module itself, I'm not sure. It should be easier for test purposes that way, at least

File: tasks/bower-release.js [hunk 7]
```diff
@@ -263,17 +271,38 @@ module.exports = function(grunt) {
         if(typeof msg !== 'string' || !msg.length)
           msg = 'Bumped version to ' + bowerJSON.version
         endpoint.commit(msg, function(err) {
-          /* Tag name must be valid semver -- but I'm not validating this here. */
-          /* TODO: Validate this here! */
-          endpoint.tag(bowerJSON.version, makeTagMsg(options.packageName), tagged)
+          if (options.overwriteVersion === true) {
```
Comment (caitp) [inline tasks/bower-release.js]: does this need to be explicitly true? is truthy not good enough?

File: tasks/endpoints/test-ep.js [hunk 8]
```diff
@@ -22,7 +22,7 @@
  * THE SOFTWARE.
  */
 
-module.exports = function(grunt) {
+function Test(grunt) {
```
Comment (caitp) [inline tasks/endpoints/test-ep.js:25]: You can annotate things to be injected into instantiated/invoked values with the `$inject` property of the function, which should be an array containing names of injectables.

I can't recall if node-di will automatically annotate based on the function signature or not. It probably does, but I can't recall. Explicit annotation might be a better idea anyways

File: test/unit/tasks/endpoints/git-test.js [hunk 9]
```diff
@@ -0,0 +1,91 @@
+'use strict';
+
```
Comment (caitp) [inline test/unit/tasks/endpoints/git-test.js:2]: +1 for including this file, I really needed a test case for the endpoint.

File: tasks/bower-release.js (hunk 5 above)
Comment (mikeborozdin) [inline tasks/bower-release.js:34]: It has always been there, see the same code on the `master` https://github.com/caitp/grunt-bower-release/blob/master/tasks/bower-release.js#L45

I do agree though that it's not how it should be. The problem is the way we test. Neither `stage_copy_test.js` nor `endpoint_calls_test` are making any calls to `bowerRelease` instead it is done in via the `test` task of `Gruntfile.js`. As you say in https://github.com/caitp/grunt-bower-release/issues/21 this is not how it should be. I wholeheartedly agree with that.

Hence, I propose to solve that problem, as well as https://github.com/caitp/grunt-bower-release/issues/20, in separate pull requests, while keeping each of them, including this one, focused.

Comment (caitp) [inline tasks/bower-release.js:34]: okay, I think when i landed that I felt it was better to have tests and worry about the infrastructure later. So I'm with you there, this can be dealt with seperately. It's unfortunate though

File: tasks/bower-release.js (hunk 7 above)
Comment (mikeborozdin) [inline tasks/bower-release.js]: Fixed

**Comment by caitp** [comment]
wouldn't worry too much about the coverage decrease, since it wasn't very good to begin with in this repo... though this PR does add some code so it would be good to sort that out

**Comment by mikeborozdin** [comment]
Thanks! I've been just looking at it right now. I suppose we'll add more PRs for introducing more robust tests, anyway.

File: tasks/endpoints/test-ep.js (hunk 8 above)
Comment (mikeborozdin) [inline tasks/endpoints/test-ep.js:25]: Done

**Comment by mikeborozdin** [comment]
Hi @caitp,

I have introduced the changes you suggested. Let me know if there's anything else.

File: tasks/bower-release.js (hunk 7 above)
Comment (caitp) [inline tasks/bower-release.js]: I meant just `if (options.overwriteVersion) {`

File: tasks/endpoints/git.js [hunk 10]
```diff
@@ -22,14 +22,14 @@
...
-module.exports = function(grunt) {
+function Git(grunt, async) {
   /* Verbosity */
   var streams = [undefined, undefined, undefined]
   if(grunt.option('verbose')) {
     streams[1] = process.stdout
     streams[2] = process.stderr
   }
-  return {
+  var objToReturn = {
```
Comment (caitp) [inline tasks/endpoints/git.js]: this object isn't being touched after declaration, so this isn't needed

**Comment by caitp** [comment]
Just a few extra nits, if you don't fix them i'll amend commits for them... I'll give this another look in a few hours, but it's probably fine for checkin.

File: tasks/endpoints/git.js (hunk 10 above)
Comment (mikeborozdin) [inline tasks/endpoints/git.js]: Indeed. I am not modifying it. However, I'm referencing a method on the same object, see [line 134](https://github.com/NCR-CoDE/grunt-bower-release/blob/feature/gbr-35/tasks/endpoints/git.js#L134)

**Comment by mikeborozdin** [comment]
Hi,

Thanks for the review! I've fixed those bits, as well, as refactored my unit tests.

File: tasks/bower-release.js (hunk 7 above)
Comment (mikeborozdin) [inline tasks/bower-release.js]: Fixed.

File: tasks/endpoints/git.js [hunk 11]
```diff
@@ -104,6 +104,47 @@ module.exports = function(grunt) {
...
+    getVersionTags: function(version, done) {
+      grunt.verbose.writeln('git tag');
+
+      grunt.util.spawn({
+        cmd: 'git',
+        args: ['tag']
+      }, function(error, result) {
+        var tags = result.stdout.split(/\n/);
+        done(tags.filter(function(tag) {
+          return tag.indexOf(version) === 0;
```
Comment (caitp) [inline tasks/endpoints/git.js:121]: this will potentially return multiple tags, for instance https://github.com/angular/bower-angular/releases could all be prefixed with an identical version, and all be removed in `removeVersionTags`

File: tasks/endpoints/git.js [hunk 12]
```diff
@@ -104,6 +104,47 @@ module.exports = function(grunt) {
...
+
+    removeLocalTag: function(tag, done) {
+      var args = ['tag', '-d', tag];
+      grunt.verbose.writeln('git ' + args.join(' '));
+      grunt.util.spawn({
+        cmd: 'git',
+        args: args,
+        opts: { stdio: streams }
+      }, function () {
+        objToReturn.removeRemoteTag(tag, done);
```
Comment (caitp) [inline tasks/endpoints/git.js]: we should not be calling this objToReturn --- if it's going to be a named object, it should be a meaningful name, so as to not confuse me when looking at changes to it in the future.

File: tasks/injection-module.js [hunk 13]
```diff
@@ -0,0 +1,10 @@
+'use strict';
+
+module.exports = function(grunt) {
+	var async = require('async');
```
Comment (caitp) [inline tasks/injection-module.js]: style, 2space indentation please

**Comment by caitp** [comment]
So, there are a few things I missed last time I looked at this --- I think the tag removal algorithm needs to be updated a bit, unless you can explain why I'm wrong about thinking that.

I really, really want to see this squashed into a minimal number of commits, as well.

The commits should be something like:

```
1. feat(endpoints): support overwriting existing tags if requested
2. chore(package.json): add "async" dependency
3. refactor(*): add "node-di" dependency to enabling better testing strategy
4. style(*): misc small style fixups
5. tests(*): refactor tests (all of test changes which are not related to changes 1 or 3 can live here)
```

For change 4, this would include changes to whitespace, newlines, removed comments, etc, log messages, etc. All of the specific style fixes should be noted in the commit message, but there only needs to be one, and it should be the last of them, because usually style fixes are unrelated. When you make style fixes for a commit, it's preferable that you ammend the commit instead, rather than introducing a separate change. But in this case, there are too many changes, so it would be hard to go back and fix that.

So, I think that should be the last bit of review for this particular CL, let me know if that SGTY

File: tasks/endpoints/git.js (hunk 11 above)
Comment (mikeborozdin) [inline tasks/endpoints/git.js:121]: True. Moreover, as line 109 suggests I am indeed removing multiple tags. 

However, it is easy to configure which tags are to be removed.

If `overwriteVersion = true` and the version in `bower.json` is set to `v1.2.17-build.205`, then `v1.2.17-build.205+sha1` and '`v1.2.17-build.205+sha will be removed`. In other words, all the tags whose name starts with ````v1.2.17-build.205` will be removed.

I don't foresee any problems here.

However, if `overwriteVersion = true` and the version in `bower.json` is `v1.3.0`, then `v1.3.0-build.2770+sha.2859fc4` and `v1.3.0-build.2769` will be deleted.

If you consider this behaviour (deleting tags whose name starts with `bowerJSON.version` if `overwriteVersion=true`) to be too dangerous, then I propose another approach. Given that we suffix tags with `+[timestamp]` when `overwriteVersion=true`, then we should delete only tags whose name pattern is `[bowerJSON.version]+`. For example, if `bower.json` has the version of `1.0.0-SNAPSHOT`, then we will delete the following tags:

```
1.0.0-SNAPSHOT+1
1.0.0-SNAPSHOT+2
```

but not `1.0.0-SNAPSHOT.a`

Comment (caitp) [inline tasks/endpoints/git.js:121]: This is problematic in the case of Angular, or similar projects which publish multiple tags for the same version (which is useful when allowing people to dogfood your framework). So yes, this is a problem

Comment (mikeborozdin) [inline tasks/endpoints/git.js:121]: @caitp,

Will you ever use `overwriteVersion` on a project like Angular?

Will the second approach solve the problem?

By the way, are you using `grunt-bower-plugin` while working on AngularJS?

Comment (caitp) [inline tasks/endpoints/git.js:121]: we aren't using it on angular at the moment, we have other scripts for publishing to bower. however it is used for other of my projects which can't really clobber past releases like this.

I am not sure I understand the second approach, but it sounds like it's still a "clobber multiple tags" thing, when the use case here is just to update one specific tag.

Why not provide a second option for clobbering all similar tags instead?

Comment (mikeborozdin) [inline tasks/endpoints/git.js:121]: The second approach is different because it only removes similar tags which are suffixed with `+`, well there can be an additional part after the plus character.

Unfortunately, we have to remove all the tags which correspond to a given version, when overwriting it, otherwise `bower update` will not fetch an overwritten version.

So, in terms of providing the second configuration option for clobbering all similar tags, how will it differ from `overwriteVersion`?

Basically, `overwriteVersion` result in the following behaviour:
1. Your Git tag will be suffixed with `+[timestamp]`
2. All 'similar' tags, i.e. whose name starts with `bowerJSON.version` will be deleted (see above why).

Comment (caitp) [inline tasks/endpoints/git.js:121]: > Unfortunately, we have to remove all the tags which correspond to a given version, when overwriting it, otherwise bower update will not fetch an overwritten version.

If you ask for `latest` or a specific tag, it will check out the latest tag, or that specific tag

> Your Git tag will be suffixed with +[timestamp]

I don't like this, I'd prefer not to be opinionated about what poeples tags look like.

> All 'similar' tags, i.e. whose name starts with bowerJSON.version will be deleted (see above why).

This is not actually true, as explained above. If you can reproduce this, then it's arguably a bug which should be filed and fixed on bower. We should not be deleting perfectly good tags which applications may still be using, unless you explicitly ask to do it.

Comment (mikeborozdin) [inline tasks/endpoints/git.js:121]: > If you ask for latest or a specific tag, it will check out the latest tag, or that specific tag

Yes, but this is not the problem the pull request is trying to solve. As I explained in https://github.com/caitp/grunt-bower-release/issues/35, the goal is to support snapshot dependencies, e.g. in my bower.json I can specify:

```
"some-cool-library: "1.0.0-SNAPSHOT"
```

As a developer of a snapshot dependency, I want to be able to overwrite it. However, the default behaviour of `grunt-bower-release` will not permit me doing that, instead it will fail, because it won't be able to push the same tag twice. 

Also, the dependency version in the example `bower.json` is a strict one, i.e. it doesn't have any approximation, such as `~` or `^`, because we are explicitly asking for a snapshot version.

Comment (caitp) [inline tasks/endpoints/git.js:121]: arguably, overwriting snapshot'd versions is a good way to break peoples applications, so it's not necessarily good.

However, if you request a tag `1.0.0-SNAPSHOT`, bower doesn't care about other tags, it will fetch this specific one. Overwriting this one specific tag is all that should be necessary to meet this use case

Comment (mikeborozdin) [inline tasks/endpoints/git.js:121]: > arguably, overwriting snapshot'd versions is a good way to break peoples applications, so it's not necessarily good.

It is true that it can break other applications, however, if someone is using a snapshot dependency they must realise all the risks. This is how it's done in Maven.

> However, if you request a tag 1.0.0-SNAPSHOT, bower doesn't care about other tags, it will fetch this specific one. Overwriting this one specific tag is all that should be necessary to meet this use case

How would you overwrite a tag? You have to delete it first and then push a tag with the same name. However, `bower update` will get new code, as it will assume it still has the old version, because the tag name has not changed. Thus, we have add some unique suffix after the plus character.

Comment (caitp) [inline tasks/endpoints/git.js:121]: people can force an update by removing their bower_components directory. you may need to remove from cache too, but that's the price you pay for that.

This is why I think using git shas or similar is a better strategy for people, but this task shouldn't do that for them

Comment (mikeborozdin) [inline tasks/endpoints/git.js:121]: > people can force an update by removing their bower_components directory. you may need to remove from cache too, but that's the price you pay for that.

Yes. One will have to do `bower cache clean`, however, the idea was to make it more transparent for developers.

> This is why I think using git shas or similar is a better strategy for people, but this task shouldn't do that for them

Do you mean using referencing a dependency by its SHA? Well, in this case, while it is guaranteed that a dependency will never change, it will not allow us to use snapshot versions either.

Anyway, as it was mentioned above, the goal of this pull request is to allow for having snapshot dependencies. We agreed to introduce the `overwriteVersion` configuration option. If is set to `true`, then:
1.  A git tag will be in the form of `bowerJSON.version+[timestamp]`, e.g. ```1.0.0-SNAPSHOT+111111`
2. If there are already tags, whose name starts with `bowerJSON.version` (e.g. `1.0.0-SNAPSHOT+2222` they will be deleted.

If `overwriteVersion` is not present or set to `false`, then the plug-in will continue in its normal mode, that is, a tag will not be suffixed and none of the tags will be removed.

@caitp, I apologise if that behaviour wasn't clear from the very beginning. I just wonder if it can still align with the `grunt-bower-release` roadmap.

Comment (caitp) [inline tasks/endpoints/git.js:121]: Removing all similar tags is fine, if any only if it is explicitly requested, either via a command line flag, or via a task parameter (separate from `overwriteVersion` --- because semantically it is doing something very different from overwriting a specific tag)

Comment (mikeborozdin) [inline tasks/endpoints/git.js:121]: Okay, cool! But what should `overwriteVersion` do without that second parameter? Just suffix a tag with a timestamp? If so, can we give it a more elaborate name, such as `suffixTagWIthTimestamp`?

Comment (caitp) [inline tasks/endpoints/git.js:121]: I think `overwrite` should do what it says it's doing --- `overwrite`-ing a specific (just one) tag

Comment (mikeborozdin) [inline tasks/endpoints/git.js:121]: So, what exactly would you consider to be overwriting a specific tag? It is effectively a multi-stage process that consists of removing similar tags and pushing a new, albeit differently, suffixed one. Otherwise, either `git push`` will fail because of an existing tag with the same name, or`bower update/install``` will not fetch a new one, unless you clean the cache. Sorry for going through this again.

Comment (caitp) [inline tasks/endpoints/git.js:121]: Overwriting a tag would constitute:

``` js
1) removing said remote tag
2) removing said local tag
3) creating the new tag
4) pushing this tag to the endpoint
```

Now, you can take this algorithm, and optionally insert a step 0, which would clobber all of the tags matching a given pattern, if a specific command line flag or option is set.

I think this is reasonable --- `overwriteVersion` does essentially what it says it's doing (where `version` refers to a `tagged version`, specifically, and so just a single tag). The option could even be renamed to `overwriteTag`.

The second option could be called whatever you like, really, `clobberSimilarTags`, `clobberVersionTags`, or some creative name for it.

I agree that 2-tags isn't necessarily the easiest way to use this, but `overwriteVersion` is really a misleading name for something which actually destroys any number of "versions" if they start with the same prefix, so if we want it to do that, we need either a better name for it, or another option to compound it into behaving that way.

Which would you see as the preferred API?

Comment (mikeborozdin) [inline tasks/endpoints/git.js:121]: > Overwriting a tag would constitute:
> 1) removing said remote tag
> 2) removing said local tag
> 3) creating the new tag
> 4) pushing this tag to the endpoint

Will step 3 create a tag whose name exactly matches `bowerJSON.version` or can it be prefixied with a unique string, say, a timestamp? I suppose the former. In this case, may I propose an introduction of the following options?
- `overwriteTag` does exactly what you described
- `suffixTagWithTimestamp` as the name suggests, it will append ```+[timestamp]` to a tag name
- `clobberVersionTags` - will delete all version tags matching `bowerJSON.version`

Comment (caitp) [inline tasks/endpoints/git.js:121]: SGTM --- this doesn't sound unreasonable

Comment (mikeborozdin) [inline tasks/endpoints/git.js:121]: Awesome :). Just to clarify `clobberVersionTags` will use a version from `bower.json` to do its dirty job, it will not require any other options in `Gruntfile.js`. Is it fine for you?

Comment (caitp) [inline tasks/endpoints/git.js:121]: that's fine

Comment (mikeborozdin) [inline tasks/endpoints/git.js:121]: Fabulous! I'll push the changes by tomorrow :)!

**Comment by mikeborozdin** [comment]
Hi @caitp,

Please, disregard the other pull request, which I've already closed. I've managed to push squashed commits here by using the `--force` option.

**Comment by caitp** [comment]
Okay, this LGTM, mostly

**Comment by caitp** [comment]
Okay, so this is merged, but I think there are some test cases missing here, and we still need to document the new options in README.md --- I've made a note in the commit message, if you want to do that. Otherwise I'll fix it when I have some time

**Comment by mikeborozdin** [comment]
Thanks a lot :)! I can do it tomorrow, if it's fine for you.
